lib/util/xpcapng.c

```diff
@@ -0,0 +1,630 @@
+enum pcapng_opt_epb {
+	PCAPNG_OPT_EPB_FLAGS = 2,
+	PCAPNG_OPT_EPB_HASH,
+	PCAPNG_OPT_EPB_DROPCOUNT,
+
+	/* The below are suggestions based on the following pull request:
+	 *   https://github.com/pcapng/pcapng/pull/91
+	 */
+	PCAPNG_OPT_EPB_PACKETID,
+	PCAPNG_OPT_EPB_QUEUE,
+	PCAPNG_OPT_EPB_VERDICT
+};
+
+enum pcapng_epb_vedict_type {
+	PCAPNG_EPB_VEDRICT_TYPE_HARDWARE = 0,
+	PCAPNG_EPB_VEDRICT_TYPE_EBPF_TC,
+	PCAPNG_EPB_VEDRICT_TYPE_EBPF_XDP
+};
+
+/*****************************************************************************
+ * pcapng_get_option_length()
+ *****************************************************************************/
+static size_t pcapng_get_option_length(size_t len)
+{
+	return roundup(sizeof(struct pcapng_option) + len, sizeof(uint32_t));
+}
+
+/*****************************************************************************
+ * pcapng_add_option()
+ *****************************************************************************/
+static struct pcapng_option *pcapng_add_option(struct pcapng_option *opt,
+					       uint16_t type, uint16_t length,
+					       const void *data)
+{
+	if (opt == NULL)
+		return NULL;
+
+	opt->po_type = type;
+	opt->po_length = length;
+	memcpy(opt->po_data, data, length);
+
+	return (struct pcapng_option *)
+		((uint8_t *)opt + pcapng_get_option_length(length));
+}
+
+/*****************************************************************************
+ * pcapng_write_shb()
+ *****************************************************************************/
+static bool pcapng_write_shb(struct xpcapng_dumper *pd, const char *comment,
+			     const char *hardware, const char *os,
+			     const char *user_application)
+{
+	int                                 rc;
+	size_t                              shb_length;
+	struct pcapng_section_header_block *shb;
+	struct pcapng_option               *opt;
+
+	if (pd == NULL) {
+		errno = EINVAL;
+		return false;
+	}
+
+	/* First calculate the total length of the SHB. */
+	shb_length = sizeof(*shb);
+
+	if (comment)
+		shb_length += pcapng_get_option_length(strlen(comment));
+
+	if (hardware)
+		shb_length += pcapng_get_option_length(strlen(hardware));
+
+	if (os)
+		shb_length += pcapng_get_option_length(strlen(os));
+
+	if (user_application)
+		shb_length += pcapng_get_option_length(
+			strlen(user_application));
+
+	shb_length += pcapng_get_option_length(0);
+	shb_length += sizeof(uint32_t);
+
+	/* Allocate the SHB and fill it. */
+	shb = calloc(shb_length, 1);
+	if (shb == NULL) {
+		errno = ENOMEM;
+		return false;
+	}
+
+	shb->shb_block_type = PCAPNG_SECTION_BLOCK;
+	shb->shb_block_length = shb_length;
+	shb->shb_byte_order_magic = PCAPNG_BYTE_ORDER_MAGIC;
+	shb->shb_major_version = PCAPNG_MAJOR_VERSION;
+	shb->shb_minor_version = PCAPNG_MINOR_VERSION;
+	shb->shb_section_length = UINT64_MAX;
+
+	/* Add the options and block_length value */
+	opt = (struct pcapng_option *) &shb->shb_options;
+
+	if (comment)
+		opt = pcapng_add_option(opt, PCAPNG_OPT_COMMENT,
```

Comment:
I get that these can't fail because the length was calculated above, but it still makes me a bit twitchy. Is there any way we could guard against, say, someone adding a new option in this part and forgetting to add it to the length calculation?

Reply:
It will definitely trash memory if they forget to add the code above for length. But the good part is that it will mess up the length in the structure and the first try to load the capture file in Wireshark or the self-test will fail as the file is invalid.

Will add a comment at the last option as a reminder to update length calculation above :)